cilium-cli/connectivity/check/fqdn.go

```diff
@@ -0,0 +1,46 @@
+// SPDX-License-Identifier: Apache-2.0
+// Copyright Authors of Cilium
+
+package check
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/cilium/cilium/cilium-cli/defaults"
+)
+
+// WithCleanFqdnCache attaches a per-scenario setup callback to the test
+// which cleans FQDN cache.
+func (t *Test) WithCleanFqdnCache() *Test {
+	t.WithScenarioSetupFunc(cleanFqdnCache)
+	return t
+}
+
+// CleanFqdnCache cleans FQDN cache of the given node via its agent pod.
+// The function signature matches SetupFunc.
+func cleanFqdnCache(ctx context.Context, t *Test, ct *ConnectivityTest, s Scenario) error {
+	for _, cp := range ct.ciliumPods {
+		if err := cleanFqdnCacheOnAgent(ctx, ct, cp); err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func cleanFqdnCacheOnAgent(ctx context.Context, log Logger, agent Pod) error {
+	log.Debugf("[%s] Cleaning FQDN cache on %s...",
+		agent.K8sClient.ClusterName(), agent.Name())
+
+	ctx, cancel := context.WithTimeout(ctx, ShortTimeout)
+	defer cancel()
+
+	if _, err := agent.K8sClient.ExecInPod(ctx, agent.Namespace(), agent.NameWithoutNamespace(),
+		defaults.AgentContainerName, []string{"cilium", "fqdn", "cache", "clean", "-f"}); err == nil {
```

Comment:
This is not safe to run in a production environment. I don't think there's a way to integrate this into this testsuite safely.

Reply:
I would be tempted to say that we could integrate this into the github workflow files, but actually that will just mean that the CI tests pass but the same problem could impact users at runtime when they attempt to run the tests to validate their own environments.

So that leads me to think that I just don't fully understand why existing entries in the FQDN cache could lead to test failures. Perhaps you could expand on the circumstances that lead to this kind of failure?

Reply:
I stumbled on this problem while working on a `connectivity test` item for #34024 (enable DNS proxy for nodes). I was getting inconsistent results running the whole test suite, my working tests, and the new `connectivity test` item by itself, and cleaning FQDN cache between individual test runs appeared to help. However, after merging my changes into the current main branch, I am unable to reproduce this behavior anymore, so I will close this PR for now. Also if I understand the code correctly, DNS proxy does not satisfy DNS _requests_ from FQDN cache - FQDN cache is used inside Cilium to map FQDN-based to IP-based policies. I still suspect that cleaning caches between tests is a good idea, however, and I don't understand how cleaning a cache can be unsafe (it's only a cache, it is only supposed to contain data which can be retrieved from elsewhere).

Reply:
The cache is populated from the user applications performing DNS requests, and that cache is used to inform the policy engine how to allow traffic corresponding to the DNS names. Losing the cached DNS names can mean that we stop allowing active connections and require the application to re-query the DNS name to provide continuity of traffic forwarding for connections it has already established. With well-behaved apps in the long term this will restore connectivity, but it also has the potential to create unnecessary disruption to active traffic.

Reply:
Speaking of traffic forwarding, I have been observing that Cilium policy lags behind Kubernetes, i.e. when a new pod comes up and starts generating traffic, I am sometimes seeing a burst of audit notifications for connections which are allowed by existing `Cilium(Clusterwide)NetworkPolicy` objects. If I switch on policy enforcement, these connections will be dropped. Is there a way to delay pod startup(?) until its Cilium endpoint is set up? (sorry if this is not the proper place to ask this question)

Reply:
Ah interesting problem area... I would suggest spinning out a dedicated issue for this for further discussion. For the node itself if the security identity is already known there should be a delay already, but there is also some cross-cluster distribution of state that needs to occur, and that can introduce such race conditions. I would also suggest including details about the size of the cluster and whether you're using CiliumEndpointSlices or not.